src/middleware/rewriteUrls.coffee

```diff
@@ -0,0 +1,104 @@
+url = require 'url'
+winston = require 'winston'
+Q = require 'q'
+
+Channel = (require '../../lib/model/channels').Channel
+config = require '../config/config'
+routerConf = config.get 'router'
+utils = require '../../lib/utils'
+router = require '../../lib/middleware/router'
+
+invertPathTransform = (pathTransform) ->
+  return pathTransform.replace /s\/(.*?)\/(.*?)($|\/(.*)$)/, 's/$2/$1/$4'
```

Comment:
Yeah this is confusing, so the inverting of the transform is only for 'virtual' config where we are rewriting a url for hosts we are proxying. In that case they may already have a pathTransform on the route that alters the path that they send a request to. So, when we get a link back from them in the response we need to do the reverse of that transform so that the link reflect the original path that the HIM channel knows about.

For example, say we have a channel with `urlPattern=/CSD/` and a pathTransform on the primary route as follows `pathTransform=s/CSD/ihris/` (ie. the actual server we are proxying is running on `http://<host>:<port>/ihirs/`). If we get links back from this server it will be something like `http://<host>:<port>/ihirs/something/123` but we need it to be `http://<him_host>:<him_port>/CSD/something/123`. To do this we can reverse (invert?) the pathTransform on the route (s/ihris/CSD/) and apply it while doing the rewrite (confusing I know).

This doesn't always works (eg. `s/CSD//`) but in that case a user can create a static rule for rewriting that will overwrite the 'virtual' rule because it is located first in the array (we only apply the first rule found).

I think I need to make the code a bit more readable, I'l add in some comments to help make this clear. Its pretty convoluted right now.